libs/pdi-scanner/src/rusb_async/mod.rs

```diff
@@ -0,0 +1,324 @@
+//     without limitation the rights to use, copy, modify, merge, publish,
+//     distribute, sublicense, and/or sell copies of the Software, and to
+//     permit persons to whom the Software is furnished to do so, subject to
+//     the following conditions:
+//
+//     The above copyright notice and this permission notice shall be
+//     included in all copies or substantial portions of the Software.
+//
+//     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+//     EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+//     MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+//     NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+//     LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+//     OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+//     WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+//
+// Modifications have been made to better support the use case of this project.
+
+use rusb::ffi;
+
+use std::convert::TryInto;
+use std::ptr::NonNull;
+
+use std::sync::atomic::{AtomicBool, Ordering};
+
+mod error;
+mod pool;
+
+pub use error::{Error, Result};
+pub use pool::TransferPool;
+
+struct Transfer {
+    ptr: NonNull<ffi::libusb_transfer>,
+    buffer: Vec<u8>,
+}
+
+impl Transfer {
+    // Invariant: Caller must ensure `device` outlives this transfer
+    #[allow(unused)]
+    unsafe fn bulk(
+        device: *mut ffi::libusb_device_handle,
+        endpoint: u8,
+        mut buffer: Vec<u8>,
+    ) -> Self {
+        // non-isochronous endpoints (e.g. control, bulk, interrupt) specify a value of 0
+        // This is step 1 of async API
+
+        let ptr =
+            NonNull::new(ffi::libusb_alloc_transfer(0)).expect("Could not allocate transfer!");
+
+        let user_data = Box::into_raw(Box::new(AtomicBool::new(false))).cast::<libc::c_void>();
+
+        let length = if endpoint & ffi::constants::LIBUSB_ENDPOINT_DIR_MASK
+            == ffi::constants::LIBUSB_ENDPOINT_OUT
+        {
+            // for OUT endpoints: the currently valid data in the buffer
+            buffer.len()
+        } else {
+            // for IN endpoints: the full capacity
+            buffer.capacity()
+        }
+        .try_into()
+        .unwrap();
+
+        ffi::libusb_fill_bulk_transfer(
+            ptr.as_ptr(),
+            device,
+            endpoint,
+            buffer.as_mut_ptr(),
+            length,
+            Self::transfer_cb,
+            user_data,
+            0,
+        );
+
+        Self { ptr, buffer }
+    }
+
+    // Invariant: Caller must ensure `device` outlives this transfer
+    #[allow(unused)]
+    unsafe fn control(
+        device: *mut ffi::libusb_device_handle,
+
+        request_type: u8,
+        request: u8,
+        value: u16,
+        index: u16,
+        data: &[u8],
+    ) -> Self {
+        let mut buf = Vec::with_capacity(data.len() + ffi::constants::LIBUSB_CONTROL_SETUP_SIZE);
```

Comment:
Rather than importing all the constants individually I just reference them through the namespace.